alicloud/service_alicloud_ecs.go

```diff
@@ -900,44 +900,49 @@ func (s *EcsService) QueryInstanceAllDisks(id string) ([]string, error) {
 	return ids, nil
 }
 
-func (s *EcsService) WaitForSnapshot(snapshotId string, status Status, timeout int) error {
+func (s *EcsService) WaitForSnapshot(id string, status Status, timeout int) error {
 	deadline := time.Now().Add(time.Duration(timeout) * time.Second)
 	for {
-		if time.Now().After(deadline) {
-			return WrapErrorf(Error(GetTimeoutMessage("Snapshot", "timeout")), DefaultTimeoutMsg, snapshotId, "WaitForSnapshot", ProviderERROR)
-		}
-		snapshot, err := s.DescribeSnapshotById(snapshotId)
+		object, err := s.DescribeSnapshot(id)
 		if err != nil {
-			if NotFoundError(err) && status == Deleted {
-				return nil
+			if NotFoundError(err) {
+				if status == Deleted {
+					return nil
+				}
 			}
-			return WrapErrorf(err, DefaultErrorMsg, snapshotId, "WaitForSnapshot", AlibabaCloudSdkGoERROR)
+			return WrapError(err)
 		}
-		if snapshot.Status == string(status) {
+
+		if object.Status == string(status) {
 			return nil
 		}
-		if snapshot.Status == string(SnapshotCreatingFailed) {
-			return WrapErrorf(Error("create snapshot failed"), DefaultErrorMsg, snapshotId, "WaitForSnapshot", ProviderERROR)
+
+		if object.Status == string(SnapshotCreatingFailed) {
+			return WrapErrorf(Error("create snapshot failed"), DefaultErrorMsg, id, "WaitForSnapshot", ProviderERROR)
 		}
 
+		if time.Now().After(deadline) {
+			return WrapErrorf(err, WaitTimeoutMsg, id, GetFunc(1), timeout, object.Status, string(status), ProviderERROR)
+		}
 		time.Sleep(DefaultIntervalShort * time.Second)
+
 	}
 }
 
-func (s *EcsService) DescribeSnapshotById(snapshotId string) (*ecs.Snapshot, error) {
-	args := ecs.CreateDescribeSnapshotsRequest()
-	args.SnapshotIds = fmt.Sprintf("[\"%s\"]", snapshotId)
+func (s *EcsService) DescribeSnapshot(id string) (*ecs.Snapshot, error) {
+	request := ecs.CreateDescribeSnapshotsRequest()
+	request.SnapshotIds = fmt.Sprintf("[\"%s\"]", id)
 	raw, err := s.client.WithEcsClient(func(ecsClient *ecs.Client) (interface{}, error) {
-		return ecsClient.DescribeSnapshots(args)
+		return ecsClient.DescribeSnapshots(request)
 	})
 	if err != nil {
```

Comment:
check notfound?

Reply:
There is not exist  error code “notfound”  filter by snapshotId